### PR TITLE
Commenting pickles

### PR DESCRIPTION
**Pull request by dannywillems**
This MR aims to commenting pickles.

Pickles makes abstraction of multiple cryptographic and type theory concepts. I took the liberty to redirect the readers to online documentations and research papers instead of explaining with new words as I feel the explanations would be better by external resources.

For the reviewers: there is a bunch of `TODO` for which I do not know the answer. Feel free to comment.

For another MR(?):
- `plonk_checks`
- `gen_scalars`

### CODE REVIEW

**Comment by dannywillems** [comment]
Considering this MR as ready to be reviewed. More changes will come in another MR.

File: src/lib/pickles_types/shifted_value.ml
Comment (dannywillems) [reply src/lib/pickles_types/shifted_value.ml]: Reverted.

File: src/lib/pickles_types/shifted_value.ml
Comment (dannywillems) [reply src/lib/pickles_types/shifted_value.ml]: Reverted - commit has been removed.

File: src/lib/pickles_types/shifted_value.mli
Comment (dannywillems) [reply src/lib/pickles_types/shifted_value.mli:32]: TODO removed

File: src/lib/pickles_types/shifted_value.mli
Comment (dannywillems) [reply src/lib/pickles_types/shifted_value.mli:43]: TODO removed

**Comment by rbonichon** [comment]
!ci-build-me

**Comment by dannywillems** [comment]
!ci-build-me

File: src/lib/pickles/intf.mli
Comment (jspada) [inline src/lib/pickles/intf.mli:84]: Is this an MSM OR just a single scalar multiplication where the scalar is passed expanded already into a list of bits?

File: src/lib/pickles/intf.mli
Comment (jspada) [inline src/lib/pickles/intf.mli:165]: Maybe also note `y` is a  quadratic residue modulo `p`

File: src/lib/pickles/make_sponge.mli
Comment (jspada) [inline src/lib/pickles/make_sponge.mli:7]: May also be good to mention any assumptions about whether the object being hashed are fixed-length or variable length with any padding applied.

File: src/lib/pickles/one_hot_vector/one_hot_vector.mli
Comment (jspada) [inline src/lib/pickles/one_hot_vector/one_hot_vector.mli:40]: I like the examples. Keep doing this imho. :-)

File: src/lib/pickles/pickles.ml
Comment (jspada) [inline src/lib/pickles/pickles.ml:78]: typo: `statements`

File: src/lib/pickles/plonk_curve_ops.mli
Comment (jspada) [inline src/lib/pickles/plonk_curve_ops.mli:54]: maybe called it `identity element` instead

File: src/lib/pickles/plonk_curve_ops.mli
Comment (jspada) [inline src/lib/pickles/plonk_curve_ops.mli:75]: Really nice commenting.  Like your style

File: src/lib/pickles/step_verifier.mli
Comment (jspada) [inline src/lib/pickles/step_verifier.mli:176]: nice

File: src/lib/pickles_base/domain.mli
Comment (jspada) [inline src/lib/pickles_base/domain.mli:1]: Part of number theoretic transform?

File: src/lib/pickles_types/nat.mli
Comment (jspada) [inline src/lib/pickles_types/nat.mli:1]: This encoding could be inefficient, why is it used?

File: src/lib/pickles_types/plonk_types.mli
Comment (jspada) [inline src/lib/pickles_types/plonk_types.mli:142]: Really good how you linked to documentation here.

**Review by jspada — APPROVED** [review]
Awesome work!  So helpful to have these comments in the code!  Well done!
Spotted a few typos, made some suggestions and asked some questions.

File: src/lib/pickles_base/domain.mli
Comment (dannywillems) [reply src/lib/pickles_base/domain.mli:1]: You're right, better to use NTT instead of FFT.

**Comment by dannywillems** [comment]
On hold for later. Converting to draft.

File: src/lib/pickles/make_sponge.mli
Comment (dannywillems) [reply src/lib/pickles/make_sponge.mli:7]: Opened https://github.com/MinaProtocol/mina/issues/14192.

File: src/lib/pickles/type.mli
Comment (dannywillems) [reply src/lib/pickles/type.mli:2]: :shrug:

File: src/lib/pickles_types/nat.mli
Comment (dannywillems) [reply src/lib/pickles_types/nat.mli:1]: Opened https://github.com/MinaProtocol/mina/issues/14193

File: src/lib/pickles_types/plonk_types.mli
Comment (dannywillems) [reply src/lib/pickles_types/plonk_types.mli:130]: :)

**Comment by dannywillems** [comment]
!ci-buid-me

File: src/lib/pickles/step_main.ml
Comment (dannywillems) [reply src/lib/pickles/step_main.ml]: Removed.

File: src/lib/pickles_types/at_most.mli
Comment (dannywillems) [reply src/lib/pickles_types/at_most.mli]: Removed.

File: src/lib/pickles_types/vector.mli
Comment (dannywillems) [reply src/lib/pickles_types/vector.mli]: Updated.

File: src/lib/pickles/intf.mli
Comment (dannywillems) [reply src/lib/pickles/intf.mli]: Updated.

File: src/lib/pickles/pickles.ml
Comment (dannywillems) [reply src/lib/pickles/pickles.ml]: Fixed.

File: src/lib/pickles/README.md
Comment (dannywillems) [reply src/lib/pickles/README.md:22]: :)

File: src/lib/pickles/README.md
Comment (dannywillems) [reply src/lib/pickles/README.md]: Updated. I should rephrase it. Not fully happy atm.

File: src/lib/pickles_types/at_most.ml
Comment (dannywillems) [reply src/lib/pickles_types/at_most.ml]: Correct. Fixed.

**Comment by dannywillems** [comment]
!ci-buid-me

**Comment by dannywillems** [comment]
!ci-build-me

File: src/lib/pickles/README.md
Comment (rbonichon) [inline src/lib/pickles/README.md:20]:
```suggestion
encode runtime invariants like vector sizes, heterogeneous lists and others at
```

File: src/lib/pickles/README.md
Comment (rbonichon) [inline src/lib/pickles/README.md:24]:
```suggestion
Some types are parametrized by type-level defined naturals.
```

File: src/lib/pickles/README.md
Comment (rbonichon) [inline src/lib/pickles/README.md:28]:
```suggestion
Functions related to computations are parametrized by at least two types,
```
Use of the American spelling for homogeneity

File: src/lib/pickles/README.md
Comment (rbonichon) [inline src/lib/pickles/README.md:33]:
```suggestion
Each in-circuit encoded value has a corresponding `'a Typ.t` value, which
```

File: src/lib/pickles/README.md
Comment (rbonichon) [inline src/lib/pickles/README.md:36]: Why "should"? Do we "need to"? or even must? Or is it something that is not enforced but that is "nice to have"?

File: src/lib/pickles/cache_handle.mli
Comment (rbonichon) [inline src/lib/pickles/cache_handle.mli:1]: FWIW: the double star `**` is not necessary for the first doc-comment. It used to be the case (if I'm not mistaken) that the first comment needed to be a single star to be used in the generated HTML!

**Review by rbonichon — COMMENTED** [review]
For now, only minor comments (mostly typos) on a ongoing partial review. Will approve once I'm done

File: src/lib/pickles/cache_handle.mli
Comment (dannywillems) [reply src/lib/pickles/cache_handle.mli:1]: Ah, I didn't know.

**Review by rbonichon — APPROVED** [review]
Nice effort! 🙇 

My comments are mainly about typos. There might also be 2/3 imprecisions, and some new simple issues to open.

File: src/lib/pickles/limb_vector/make.mli
Comment (dannywillems) [reply src/lib/pickles/limb_vector/make.mli:5]: Wanted to differentiate the integer/natural that is defined by the type system and the actual integer value (i.e. "runtime" value).

File: src/lib/pickles/per_proof_witness.mli
Comment (dannywillems) [reply src/lib/pickles/per_proof_witness.mli:1]: Yep, better.

File: src/lib/pickles/requests.mli
Comment (dannywillems) [reply src/lib/pickles/requests.mli:2]: Nop, wrong punctuation :D.

File: src/lib/pickles/ro.mli
Comment (dannywillems) [inline src/lib/pickles/ro.mli:1]:
```suggestion
(** Implements an interface simulating a {{ https://en.wikipedia.org/wiki/Random_oracle } random oracle }. *)
```

File: src/lib/pickles/requests.mli
Comment (dannywillems) [inline src/lib/pickles/requests.mli:2]:
```suggestion
    non-deterministic witness data from the prover.
```

File: src/lib/pickles/ro.mli
Comment (dannywillems) [reply src/lib/pickles/ro.mli:1]: Ok, done. Trying to do one by one. I might be missing some.

File: src/lib/pickles/sponge_inputs.mli
Comment (dannywillems) [reply src/lib/pickles/sponge_inputs.mli:10]: Glad it is :).

File: src/lib/pickles_base/proofs_verified.mli
Comment (dannywillems) [reply src/lib/pickles_base/proofs_verified.mli:17]: Opened https://github.com/MinaProtocol/mina/issues/14205

Comment (dannywillems) [reply src/lib/pickles_base/proofs_verified.mli:17]: Closing the thread as it is a `Failure` atm

File: src/lib/pickles_base/proofs_verified.mli
Comment (dannywillems) [reply src/lib/pickles_base/proofs_verified.mli:21]: Opened https://github.com/MinaProtocol/mina/issues/14206.
Closing the thread as atm it is a `Failure`.

File: src/lib/pickles_types/vector.mli
Comment (dannywillems) [reply src/lib/pickles_types/vector.mli:28]: hehe :)

File: src/lib/pickles_types/hlist.mli
Comment (dannywillems) [reply src/lib/pickles_types/hlist.mli:61]: But, we should update the others then to stay consistent.

File: src/lib/pickles_types/shifted_value.mli
Comment (dannywillems) [reply src/lib/pickles_types/shifted_value.mli]: Right. Updated.

File: src/lib/pickles/type.mli
Comment (dannywillems) [reply src/lib/pickles/type.mli]: Updated.

File: src/lib/pickles/README.md
Comment (dannywillems) [reply src/lib/pickles/README.md]: I changed to must. I hope I am not wrong.

File: src/lib/pickles/plonk_checks/plonk_checks.mli
Comment (dannywillems) [reply src/lib/pickles/plonk_checks/plonk_checks.mli:3]: Opened https://github.com/MinaProtocol/mina/issues/14208.

File: src/lib/pickles_base/domains.mli
Comment (dannywillems) [reply src/lib/pickles_base/domains.mli:2]: Opened https://github.com/MinaProtocol/mina/issues/14209.

**Comment by dannywillems** [comment]
!ci-build-me

**Comment by dannywillems** [comment]
!ci-build-me

**Comment by dannywillems** [comment]
!ci-build-me

**Pull request merged**
